Objects/typeobject.c

```diff
@@ -1075,7 +1177,7 @@ type_modified_unlocked(PyTypeObject *type)
        We don't assign new version tags eagerly, but only as
        needed.
      */
-    ASSERT_TYPE_LOCK_HELD();
+    ASSERT_NEW_TYPE_OR_LOCKED(type);
```

Comment:
```suggestion
    ASSERT_NEW_TYPE_OR_LOCKED(type);
    assert(!types_world_is_stopped());
```
Since this is re-entrant, the world shouldn't be stopped.